src/pydicom/dataset.py

```diff
@@ -1633,9 +1656,16 @@ def convert_pixel_data(self, handler_name: str = "") -> None:
         # Check if already have converted to a NumPy array
         # Also check if pixel data has changed. If so, get new NumPy array
         already_have = True
+
         if not hasattr(self, "_pixel_array"):
             already_have = False
-        elif self._pixel_id != get_image_pixel_ids(self):
+        elif self._pixel_array is None:
+            already_have = False
+
+        # The _pixel_id may give a false negative if the pixel data memory
+        #   has been freed prior to setting a new value it may reuse that memory
+        #   and therefore give the same `id()` value
```

Comment:
I had a hard time deciphering this comment until I relaized it actually is related to the change above (e.g. the check for `_pixel_array`), but I guess that is just me being slow.

Reply:
The comment confused me a bit too, mainly because I think it should be 'false positive' - i.e. the 'id' does match, when it shouldn't

Oh, and you could combine the two conditions together, especially with a `getattr` with default `None` to make the logic more clear.

Reply:
The problem is that:
```python
if not getattr(self, "_pixel_array", None):
    already_have = False
``` 
Raises a `ValueError` when the result is an `ndarray` because "The truth value of an array with more than one element is ambiguous".